lib/graphql/schema/build_from_ast.rb

```diff
@@ -0,0 +1,260 @@
+      # @return [GraphQL::Schema] the schema described by `document`
+      def build(document)
+        raise InvalidDocument.new('Must provide a document ast.') if !document || !document.is_a?(GraphQL::Language::Nodes::Document)
+
+        schema_definition = nil
+        types = {}
+        directives = {}
+        type_resolver = -> (type) { -> { resolve_type(types, type) } }
+
+        document.definitions.each do |definition|
+          case definition
+          when GraphQL::Language::Nodes::SchemaDefinition
+            raise InvalidDocument.new('Must provide only one schema definition.') if schema_definition
+            schema_definition = definition
+          when GraphQL::Language::Nodes::EnumTypeDefinition
+            types[definition.name] = build_enum_type(definition, type_resolver)
+          when GraphQL::Language::Nodes::ObjectTypeDefinition
+            types[definition.name] = build_object_type(definition, type_resolver)
+          when GraphQL::Language::Nodes::InterfaceTypeDefinition
+            types[definition.name] = build_interface_type(definition, type_resolver)
+          when GraphQL::Language::Nodes::UnionTypeDefinition
+            types[definition.name] = build_union_type(definition, type_resolver)
+          when GraphQL::Language::Nodes::ScalarTypeDefinition
+            types[definition.name] = build_scalar_type(definition, type_resolver)
+          when GraphQL::Language::Nodes::InputObjectTypeDefinition
+            types[definition.name] = build_input_object_type(definition, type_resolver)
+          when GraphQL::Language::Nodes::DirectiveDefinition
+            directives[definition.name] = build_directive(definition, type_resolver)
+          end
+        end
+
+        GraphQL::Schema::DIRECTIVES.each do |built_in_directive|
+          directives[built_in_directive.name] = built_in_directive unless directives[built_in_directive.name]
+        end
+
+        if schema_definition
+          if schema_definition.query
+            raise InvalidDocument.new("Specified query type \"#{schema_definition.query}\" not found in document.") unless types[schema_definition.query]
+            query_root_type = types[schema_definition.query]
+          end
+
+          if schema_definition.mutation
+            raise InvalidDocument.new("Specified mutation type \"#{schema_definition.mutation}\" not found in document.") unless types[schema_definition.mutation]
+            mutation_root_type = types[schema_definition.mutation]
+          end
+
+          if schema_definition.subscription
+            raise InvalidDocument.new("Specified subscription type \"#{schema_definition.subscription}\" not found in document.") unless types[schema_definition.subscription]
+            subscription_root_type = types[schema_definition.subscription]
+          end
+        else
+          query_root_type = types['Query'] if types['Query']
```

Comment:
What do we gain by these `if types[...]` suffixes? would it be the same to "just" get `nil` by a hash lookup miss, eg 

```ruby 
# query_root_type = types['Query'] if types['Query']
# is that equivalent to?
query_root_type = types['Query'] 
```